flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java

```diff
@@ -265,10 +265,15 @@ public static TaskManagerServices fromConfiguration(
 		// start the I/O manager, it will create some temp directories.
 		final IOManager ioManager = new IOManagerAsync(taskManagerServicesConfiguration.getTmpDirPaths());
 
+		final ExecutorService ioExecutor = Executors.newFixedThreadPool(
```

Comment:
Can the `ioExecutor` also replace the `taskIOExecutor`?

Reply:
Technically yes, but I wouldn't know how many threads the executor should have. It is also easier to judge how many you'd need if the task and TM operations are separated.

Reply:
I think we should really unify them. `taskIOExecutor` is effectively the same as `ioExecutor` because it is used by the `TaskExecutorLocalStateStoresManager` for discarding local state. I think it is better to use common thread pools until we have a really good reason for separating thread pools. Creating more thread pools will also increase the required resource foot print.

Concerning the number of threads, I would suggest to use `ClusterOptions.CLUSTER_IO_EXECUTOR_POOL_SIZE` which defaults to the number of available cores if it has not been specified. This will also allow users to increase the thread pool if things take too long with a single thread.

Reply:
Well, one reason is uncertainty.

Do you know how many resources the state cleanup currently siphons from the [RpcService executor](https://github.com/apache/flink/blob/5dededa06592e4b68023bb3a3282849038f18fa2/flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java#L373)?

I don't disagree with your take in general, but as is the PR is pretty safe to back-port to older versions.
I'm not so sure about this if we include the taskIOExecutor change; we don't have to look back very far for issues when we moved things away from the RPC executor (FLINK-17248).

Reply:
I think it is actually a good idea to move it away from the `RpcService's` executor because this is the same pool which is used to run the `RpcEndpoints`.